### PR TITLE
325: Create a api which takes inputes and generate returns the blocks data and statistics for blockspan - backend

### PR DESCRIPTION
**Pull request by Natsvora**
325: Updated blockspan list API to have all required data

### CODE REVIEW

File: backend/src/oe/api/routes.ts [hunk 1]
```diff
@@ -245,27 +246,29 @@ class OpEnergyRoutes {
...
+      const withNbdr = toBoolean(String(req.query.nbdr));
 
       if (
-        !isValidNaturalNumber(startBlockHeight) ||
+        !isValidPositiveNumber(endBlockHeight) ||
         !isValidPositiveNumber(span) ||
-        !isValidPositiveNumber(numberOfSpan)
+        !Number.isInteger(+numberOfSpan) ||
+        +span < config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT ||
+        (+numberOfSpan <= 0 && +numberOfSpan !== -1)
```
Comment (dambaev) [inline backend/src/oe/api/routes.ts:261]: this condition confuses me: is numberOfSpan is now allowed to be -1? What is the use case? Any number of spans until height 0?

File: backend/src/oe/api/op-blockspan.service.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import { NbdrStatisticsError } from './interfaces/op-statistics.interface';
+
+export class OpBlockSpanApiService {
+  constructor() {}
+
+  public $getBlockSpanList(UUID: string, endBlockHeight: number, span: number, blockSpanCount: number): BlockSpan[] {
+    try {
+      const blockSpanList = [] as BlockSpan[];
+      const numberOfSpan = blockSpanCount === -1 ? Number.MAX_VALUE : blockSpanCount;
+      for (let i = endBlockHeight; (i - span) > config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT && blockSpanList.length < numberOfSpan; i -= span) {
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: I see, that `i` can be negative integer: assume, that endBlockHeight = 6, span = 6 (which is current minimum) and numberOfSpan = -1, I think, that we should get 1 element in the list, but we will get Number.MAX_VALUE, which is incorrect, isn't it?

File: backend/src/oe/api/op-blockspan.service.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import { BlockSpan, BlockSpanDetails, BlockSpanDetailsWithNbdr, ConfirmedBlockHeight } from './interfaces/op-energy.interface';
+import { NbdrStatisticsError } from './interfaces/op-statistics.interface';
+
+export class OpBlockSpanApiService {
+  constructor() {}
+
+  public $getBlockSpanList(UUID: string, endBlockHeight: number, span: number, blockSpanCount: number): BlockSpan[] {
+    try {
+      const blockSpanList = [] as BlockSpan[];
+      const numberOfSpan = blockSpanCount === -1 ? Number.MAX_VALUE : blockSpanCount;
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: given a comment below, it should be something like `Math.min( Math.floor(endBlockHeight / span), blockSpanCount === -1 ? Number.MAX_VALUE : blockSpanCount)`

This way we will be able to keep `i` as Natural integer

File: backend/src/oe/api/op-blockspan.service.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    span: number,
+    blockSpanCount: number,
+    withNbdrStatistics: boolean
+  ): Promise<BlockSpanDetails[] | BlockSpanDetailsWithNbdr[] | NbdrStatisticsError> {
+    try {
+      const blockSpanDetailedList = await this.$generateBlockSpanDetailedList(
+        UUID,
+        endBlockHeight,
+        span,
+        withNbdrStatistics && blockSpanCount !== -1 ? blockSpanCount + 100 : blockSpanCount
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: this one is a bit strange: we have 3 meanings of blockSpanCount:
1. -1 means "as much spans as possible";
2. 0 is impossible;
3. other values are possible (positive numbers).

but here I see, that if I blockSpanCount = 10, you will generate list for 110 spans and I don't understand why.

File: backend/src/oe/api/op-blockspan.service.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      while (index < blockHeadersList.length - 1) {
+        result.push({
+          startBlock: {
+            height: blockHeadersList[index].height,
+            hash: blockHeadersList[index].current_block_hash,
+            mediantime: blockHeadersList[index].mediantime,
+            timestamp: blockHeadersList[index].timestamp,
+          },
+          endBlock: {
+            height: blockHeadersList[++index].height,
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: `++index` is hard to distinguish here, please, add at least spaces around it so it will be more visible

File: backend/src/oe/api/op-blockspan.service.ts (hunk 4 above)
Comment (dambaev) [reply backend/src/oe/api/op-blockspan.service.ts]: overall, given  that numberOfSpans either -1 or positive (ie, we are not allowing to have numberOfSpans = 0), I suggest to replace -1 with 0: this way we will be able to simplify input validation to `>= 0` condition

File: backend/src/oe/api/op-blockspan.service.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        span,
+        withNbdrStatistics && blockSpanCount !== -1 ? blockSpanCount + 100 : blockSpanCount
+      );
+      if (!withNbdrStatistics) {
+        return blockSpanDetailedList;
+      }
+
+      for (let index = 100; index < blockSpanDetailedList.length; index++) {
+        const blockSpanListForStat = blockSpanDetailedList.slice(index - 100, index);
+        const statistics = opStatisticsService.calculateStatisticsWithBlockSpan(blockSpanListForStat, span);
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: then, given generated list of spans of 110 elements, you will calculate statistics for [0;100] of it. Why the rest 10 will be ignored? Isn't [n - 100; n] is more preferred data here?

Comment (dambaev) [reply backend/src/oe/api/op-blockspan.service.ts]: overall, isn't 100 should be actually a `Math.min( 100, blockSpanDetailedList.length)`?

File: backend/src/oe/api/op-blockspan.service.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+          return statistics;
+        }
+        blockSpanDetailedList[index]['nbdr'] = { ...statistics.nbdr };
+      }
+
+      if (blockSpanCount === -1) {
+        return blockSpanDetailedList;
+      }
+
+      if (blockSpanDetailedList.length > 100) {
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: this condition seems always to be true: if numberOfSpans = -1, then blockSpanDetailedList.length = `Number.MAX_VALUE`, otherwise, it will be `blockSpanCount + 100`, given that numberOfSpans is Positive, this means, that numberOfSpans is always `>=101`

Comment (dambaev) [reply backend/src/oe/api/op-blockspan.service.ts]: overall statistics code is hard to follow: I can't stop thinking, that this code assumes, that there are >100 confirmed blocks in the chain. But what if tomorrow  we will deploy new instance of bitcoin node and op-energy? then this node will start chain synchronization filling everything from ground up. This will mean, that there will be 1..N of confirmed blocks. What results will be given by the current code, that is dancing around value of 100? Or what is tomorrow there will be new bitcoin network introduced (similar to signet), which will start chain from the very beginning...

File: backend/src/oe/api/op-blockspan.service.ts (hunk 2 above)
Comment (dambaev) [reply backend/src/oe/api/op-blockspan.service.ts]: okay, my bad, I missed the `(i - span) > config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT` part of condition

Comment (dambaev) [reply backend/src/oe/api/op-blockspan.service.ts]: but now I think. that `(i - span) > config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT` condition is not correct as well: we assume, that there can be a chain of confirmed blocks count = 1, ie, imagine a chain of 7 blocks, thus block 0 is assumed to be confirmed, but will it pass this condition?

Comment (dambaev) [reply backend/src/oe/api/op-blockspan.service.ts]: for me, it seems that `config.OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT` is only usable at a level of OpBlockHeaderRepository / OpBlockHeaderService as this value defines difference between bitcoin node's chain length and our's `blocks` table. What you are trying to define is `Math.floor( endBlockHeight / span)` as for me...

File: backend/src/oe/api/op-statistics.service.ts [hunk 8]
```diff
@@ -1,74 +1,46 @@
...
-          );
-
-        for (let i = 0; i < NUMBER_OF_BLOCK_SPANS; i += 1) {
-          const startBlock = blockHeadersList[i + 1];
-          const endBlock = blockHeadersList[i];
-          const nbdr =
-            (blockSpan * 600 * 100) /
-            (startBlock.timestamp - endBlock.timestamp);
+      const blockSpanList = await opBlockspanService.$generateBlockSpanDetailedList(
+        UUID, blockHeight.value - blockSpan, blockSpan, NUMBER_OF_BLOCK_SPANS
```
Comment (dambaev) [inline backend/src/oe/api/op-statistics.service.ts]: why do we have `blockHeight.value - blockSpan` here as the second argument is `endBlockHeight`?

**Review by dambaev — CHANGES_REQUESTED** [review]

File: backend/src/oe/api/routes.ts (hunk 1 above)
Comment (Natsvora) [reply backend/src/oe/api/routes.ts:261]: -1 is a special case when the user specifies -1 then we will generate the blockspan list from current verified block to zero (or near to zero)

File: backend/src/oe/api/op-statistics.service.ts (hunk 8 above)
Comment (Natsvora) [reply backend/src/oe/api/op-statistics.service.ts]: Generating statistics data we would require 100 block spans which comes before the current one. Therefore to omit the current block span from the list generated, we'll need to subtract the span from the height value.

File: backend/src/oe/api/op-statistics.service.ts [hunk 9]
```diff
@@ -1,74 +1,46 @@
...
-        logger.err(`Error while calculating nbdr ${error}`);
-        throw new Error('Error while calculating nbdr');
-      }
+  public calculateStatisticsWithBlockSpan(
+    blockSpanList: BlockSpanDetails[],
+    span: number
+  ): NbdrStatistics | NbdrStatisticsError {
+    try {
+      const nbdrStatisticsList: number[] = [];
+      blockSpanList.forEach((blockSpan) => {
```
Comment (dambaev) [inline backend/src/oe/api/op-statistics.service.ts]: by the way, it seems, that mean should be:
```
const mean =
        nbdrStatisticsList.reduce((a, b) => a + b) / blockSpanList.length;
```
now, isn't it?

**Review by dambaev — DISMISSED** [review]

File: backend/src/oe/api/op-statistics.service.ts (hunk 9 above)
Comment (Natsvora) [reply backend/src/oe/api/op-statistics.service.ts]: blockSpanList.length and NUMBER_OF_BLOCK_SPANS both have same value 100. Let me change it to blockSpanList.length

File: backend/src/oe/api/op-blockspan.service.ts (hunk 4 above)
Comment (Natsvora) [reply backend/src/oe/api/op-blockspan.service.ts]: I suggest not using 0 in any case. Else in the future, it will create confusion for us also. Even if anyhow user enters 0 we will be sure that we have handled that scenario. So let's keep -1 for all blocks and 0 as not allowed

File: backend/src/oe/api/op-blockspan.service.ts [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        blockSpanCount
+      );
+      const confirmedBlockSpanList = await this.$getConfirmedBlockSpanList(UUID, blockSpanList);
+      if (!confirmedBlockSpanList.length) {
+        return [];
+      }
+
+      const blockHeadersList = await opBlockHeaderService.$getBlockHeadersByHeights(
+        UUID,
+        Array.from(new Set(confirmedBlockSpanList))
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: I guess, that expression's `Array.from(new Set(confirmedBlockSpanList))` goal is to provide list of  unique elements? But as far as I can evaluate `$getConfirmedBlockSpanList` is not able to provide list of duplicate elements.

File: backend/src/oe/api/op-blockspan.service.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    }
+  }
+
+  public async $getBlockSpanDetailedList(
+    UUID: string,
+    endBlockHeight: number,
+    span: number,
+    blockSpanCount: number,
+    withNbdrStatistics: boolean
+  ): Promise<BlockSpanDetails[] | BlockSpanDetailsWithNbdr[] | NbdrStatisticsError> {
```
Comment (dambaev) [inline backend/src/oe/api/op-blockspan.service.ts]: Today I tried to evaluate this function for a second time and came to conclusion, that I have hard time of understanding what am I expecting as the result: maybe function's comment and example result will be useful. Because what I expect now based on function's description by calling function
```
$getBlockSpanDetailedList( UUID, 100, 1, 100, true) = {
  'list': [ { startBlock: 99, endBlock: 100}, ... { startBlock: 0, endBlock: 1}],
  'statistics': { // here goes nbdr statistics for range  [0 .. 100]
     mean: someMean,
     stdavg: someAvg
  }
}
```

But instead, I see that I will get:
```
[ (list of 100 elements of type  BlockSpanDetails without statistics) ] + [ (list of 100 elements of type BlockSpanDetail and statistics)]
```
such results confuses me a lot and I don't see explanations in the comments for example of why I don't see result expected by me but I will get result, that consist of 2 lists: one without statistics and another with statistic. So I don't understand how to treat such result at the end

**Review by dambaev — CHANGES_REQUESTED** [review]

File: backend/src/oe/api/op-blockspan.service.ts (hunk 2 above)
Comment (tphyahoo) [reply backend/src/oe/api/op-blockspan.service.ts]: re OP_ENERGY.CONFIRMED_BLOCKS_AMOUNT I think this config variable shouldn't even be available to the code. 

the only component that should know CONFIRMED_BLOCKS_AMOUNT is oe-BlockHeaderTableInserter service

what am I missing here?

File: backend/src/oe/api/op-blockspan.service.ts [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        };
+        blockSpanList.push(blockSpan);
+      }
+      return blockSpanList;
+    } catch (e) {
+      throw new Error(`${UUID} OpBlockSpanApiService.$getBlockSpanList: error while generating block span list: ${e instanceof Error ? e.message : e}`);
+    }
+  }
+
+  private async $getConfirmedBlockSpanList(UUID: string, blockSpanList: BlockSpan[]): Promise<ConfirmedBlockHeight[]> {
```
Comment (tphyahoo) [inline backend/src/oe/api/op-blockspan.service.ts]: something looks wrong to me. I don't think this method getConfirmedBlockSpanList should even exist.

The api shouldn't have any concept of if a block is confirmed or not. 
If it is in the new Blocks table, it is confirmed. Period. 

We trust the blockHeaderService to take care of this. 

So this code is greatly overcomplicated.

File: backend/src/oe/api/op-blockspan.service.ts (hunk 10 above)
Comment (tphyahoo) [reply backend/src/oe/api/op-blockspan.service.ts]: -when I look at "files changed" view for this pull request and do search for confirm I see 27 hits across a number of files, and most of them are in green areas meaning we are adding logic to do something with whether a block is confirmed.

If I understand the new data flow, This Should Never Happen.
If it's in the new blocks table, it's confirmed. 

Alex, do you agree with me here? Or am I missing something? 

Naimish, assuming alex agrees with me, I think you are misunderstanding something important here.

File: backend/src/oe/service/op-block-header.service.ts [hunk 13]
```diff
@@ -145,6 +145,11 @@ export class OpBlockHeaderService {
     }
     return { value: blockHeight };
   }
+
+  public isBlockHeightVerified(blockHeight: number, currentTip: number): boolean {
```
Comment (tphyahoo) [inline backend/src/oe/service/op-block-header.service.ts]: I think adding this method is a mistake. 
it definitely shouldn't be public. 

it's part of the mess of exposing whether a block is confirmed / verified. 
this shouldn't leak into other modules. 
(and I don't think it is required privately either.)

File: backend/src/oe/api/interfaces/op-energy.interface.ts [hunk 14]
```diff
@@ -106,3 +107,17 @@ export interface RegisterResponse {
   accountToken: string;
   accountSecret: string;
 }
+
+export interface BlockDetails {
+  height: number,
+  hash: string,
```
Comment (girish-velas) [inline backend/src/oe/api/interfaces/op-energy.interface.ts:113]: sdflksjdf

File: backend/src/oe/api/op-blockspan.service.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+import bitcoinApi from '../../api/bitcoin/bitcoin-api-factory';
+import opStatisticsService from './op-statistics.service';
+
+import { BlockSpan, BlockSpanDetails, BlockSpanDetailsWithNbdr, ConfirmedBlockHeight } from './interfaces/op-energy.interface';
+import { NbdrStatisticsError } from './interfaces/op-statistics.interface';
+
+export class OpBlockSpanApiService {
+  constructor() {}
+
+  public $getBlockSpanList(UUID: string, endBlockHeight: number, span: number, blockSpanCount: number): BlockSpan[] {
```
Comment (girish-velas) [inline backend/src/oe/api/op-blockspan.service.ts]: slkdjfslkjfsd

File: backend/src/oe/api/op-blockspan.service.ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        blockSpanCount === -1 ? Number.MAX_VALUE : blockSpanCount
+      );
+      for (let i = endBlockHeight; blockSpanList.length < numberOfSpan; i -= span) {
+        const blockSpan = {
+          startBlockHeight: i - span,
+          endBlockHeight: i
+        };
+        blockSpanList.push(blockSpan);
+      }
+      return blockSpanList;
```
Comment (girish-velas) [inline backend/src/oe/api/op-blockspan.service.ts]: sdfsdf

File: backend/src/oe/api/interfaces/op-energy.interface.ts [hunk 17]
```diff
@@ -106,3 +107,17 @@ export interface RegisterResponse {
...
+  mediantime: number,
+  timestamp: number,
+}
+
+export interface BlockSpanDetails {
+  startBlock: BlockDetails,
```
Comment (girish-velas) [inline backend/src/oe/api/interfaces/op-energy.interface.ts:119]: asdfsad

**Review by girish-velas — CHANGES_REQUESTED** [review]

**Review by tphyahoo — CHANGES_REQUESTED** [review]

**Comment by Natsvora** [comment]
https://github.com/dambaev/op-energy/pull/330#discussion_r1116077770

@tphyahoo Regarding your above comment. No, we are not doing any changes in the confirmed blocks table. 

We are just trying to fetch data from the blocks table.

File: backend/src/oe/api/oe-blockspan.service.ts [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+  > {
+    try {
+      const blockSpanDetailedList = await this.$generateBlockSpanDetailedList(
+        UUID,
+        endBlockHeight,
+        span,
+        blockSpanCount
+      );
+      return blockSpanDetailedList;
+      // TODO: Add nbdr statistics data to the response payload if withNbdrStatistics is true
```
Comment (girish-velas) [inline backend/src/oe/api/oe-blockspan.service.ts:147]: should this TODO be addressed in this PR itself?

File: backend/src/oe/api/routes.ts [hunk 19]
```diff
@@ -29,7 +30,7 @@ class OpEnergyRoutes {
       .post(config.MEMPOOL.API_URL_PREFIX + 'user/displayname', this.$postUserDisplayName)
       .get(config.MEMPOOL.API_URL_PREFIX + 'statistics/:blockheight/:span', this.$getBlockSpanStatistics)
       .get(config.MEMPOOL.API_URL_PREFIX + 'oe/block/:hash', this.$getBlockByHash)
-      .get(config.MEMPOOL.API_URL_PREFIX + 'oe/blockspanlist/:startBlockHeight/:span/:numberOfSpan', this.$getBlockSpanList)
+      .get(config.MEMPOOL.API_URL_PREFIX + 'oe/blockspanlist/:endBlockHeight/:span/:numberOfSpan', this.$getBlockSpanDetailedList)
```
Comment (girish-velas) [inline backend/src/oe/api/routes.ts:33]: why are we changing the startBlockHeight to endBlockHeight?

File: backend/src/oe/api/oe-blockspan.service.ts [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+    span: number,
+    blockSpanCount: number
+  ): Promise<BlockSpan[]> {
+    try {
+      const blockSpanList = [] as BlockSpan[];
+      const numberOfSpan = Math.min(
+        Math.floor(endBlockHeight / span),
+        blockSpanCount === -1 ? Number.MAX_VALUE : blockSpanCount
+      );
+      const currentTip = await bitcoinApi.$getBlockHeightTip();
```
Comment (girish-velas) [inline backend/src/oe/api/oe-blockspan.service.ts:26]: we shouldn't call bitcoin api from this service
